app/controllers/admin/base_admin_controller.rb

```diff
@@ -1,11 +1,25 @@
 module Admin
   class BaseAdminController < ApplicationController
-    before_action :ensure_authenticated_user
+    TIMEOUT_LENGTH_IN_MINUTES = 30
+
+    before_action :ensure_authenticated_user, :end_expired_sessions, :update_last_seen_at
 
     private
 
     def ensure_authenticated_user
       redirect_to admin_sign_in_path unless signed_in?
     end
+
+    def end_expired_sessions
+      if admin_session_timed_out?
+        session.destroy
+        redirect_to admin_sign_in_path, notice: "Your session has timed out due to inactivity, sign-in again"
```

Comment:
I'd probably add a 'please' to the 'sign-in again' instruction, just to be nicer 😄

Reply:
👍